Container Images: Collect container images from Openshift

enoodle opened the pull request:
This will collect all the images known to Openshift in addition to images that are run on pods.

It supposes not to show duplicated images, but for now until kubernetes will report the pullable imageID there is no good way to identify image versions.

![openshift_image_details](https://cloud.githubusercontent.com/assets/3123328/20561490/91eab05c-b187-11e6-8238-0d80a652e755.png)

Based on #12711

enoodle commented:
@miq-bot wip providers/containers enhancement

enoodle commented:
cc @simon3z @zeari

zeari commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -24,8 +24,6 @@ def ems_inv_to_hashes(inventory)
       get_services(inventory)
       get_component_statuses(inventory)
-      get_registries
-      get_images
```
We dont want these on Kubernetes anymore?

zeari commented:
There should probably be spec tests for duplicate images - images from both the image string and actual image entities.

enoodle commented:
@zeari I am not sure what counts as "duplicate". I sometimes have the same image with the same tag but different hash, does it count as duplicate?

![openshift_duplicate_images](https://cloud.githubusercontent.com/assets/3123328/17889065/0c45ed60-6937-11e6-8834-9e89f70f4d8c.png)

enoodle commented:
ping @simon3z what do you think about the different image versions that this collects? (see the picture in the previous comment https://github.com/ManageIQ/manageiq/pull/10692#issuecomment-241693006 )
Do you think we should collect all of those images? I am not sure about those that we have with the `latest` tag and we have them twice with and without the registry and with different hash.

simon3z commented:
@zeari please review.

> Do you think we should collect all of those images? I am not sure about those that we have with the latest tag and we have them twice with and without the registry and with different hash.

@enoodle yes they should be collected multiple times if the hash is different.

@enoodle also I don't see where you're merging the ones coming from the Pods (and the logic on how you match them).

enoodle commented:
@simon3z I was supposed to be merged by the hash using the `:by_hash_and_registry_host_port` hash, but then I re-added the `get_images` function to kubernetes and it didn't work anymore.
I have now added a test for to assert that duplicated images are not present and also implemented the merge (using the `uniq!` function ) which depends on `image_ref`.

enoodle commented:
@miq-bot remove_label wip

simon3z commented:
I have the feeling that @zakiva produced specific tests for part of the refresh (e.g. persistent volumes). Could you check those and evaluate if we can have more coverage rather than just the VCR? Thanks.

yaacov commented on spec/models/manageiq/providers/kubernetes/container_manager/metrics_capture/capture_context_spec.rb:
```diff
@@ -26,7 +26,8 @@
 
-    VCR.use_cassette("#{described_class.name.underscore}_refresh") do # , :record => :new_episodes) do
+    VCR.use_cassette("#{described_class.name.underscore}_refresh",
+                     :match_requests_on => [:path,]) do # , :record => :new_episodes) do
```
Can we remove the `match_requests_on` once the recording is done and we use the cassette ?

enoodle replied:
I think its better to stay there, this way we won't be depended on the host name.

yaacov replied:
OK :+1:

simon3z commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -109,6 +119,17 @@ def parse_build_pod(build_pod)
+        name = docker_json["name"]
+        tag = docker_json["tag"]
+        image_name = "#{name}:#{tag}"
+        image_id = "docker://%s" % openshift_image[:metadata][:name]
```
@enoodle is it really that the ID? I thought it was the hash of the image.

enoodle replied:
Yes, The imageID looks like that in the container definition that we used so far too[1]
Upon more examination I understand now that they have many different "ID"s and names there and I want to look at it again to find the right one...

@miq-bot add_label wip

https://gist.github.com/enoodle/acbdc49b73804dec3004f9c34bf619fa#file-gistfile1-txt-L135

enoodle commented:
@miq-bot add_label wip

simon3z commented:
@miq-bot assign enoodle

simon3z commented:
Blocked on OpenShift https://trello.com/c/HOWz6ejY

zeari commented:
ping @enoodle 
We would like to have to persist real openshift entities in the near future.
cc @simon3z

enoodle commented:
@zeari It is almost ready, I will soon post what I have. But before this can be merged we will need a new version of kubeclient with the discovery support (to replace the current openshift_client) to be merged to ManageIQ.

enoodle commented:
@zeari @simon3z I am sorry, I have made a mistake it isn't dependent on other PRs.
I was using the `parse_base_item` function to get the basic attributes from `metadata` but it seems that `ContainerImage` never had some of those attributes (`ems_ref`, `ems_created_on`, `namespace`, `resource_version`).
@zeari are those attributes important for you ?

zeari commented:
> @zeari are those attributes important for you ?

Not currently. 
@simon3z I guess `ems_created_on` would be the usefull one but i dont think we need those.

simon3z commented:
> Not currently. 
> @simon3z I guess ems_created_on would be the usefull one but i dont think we need those.

@enoodle @zeari both `ems_ref` and `ems_created_on` are useful but I guess we'll have to do without them now.

`ems_ref` is the unique id of the entity within the provider. I guess in our case it would have been the `metadata.uid` or `metadata.name` of the Image object in OpenShift.

```
kind: Image
metadata:
  creationTimestamp: 2016-06-01T12:52:42Z
  name: sha256:048814904e6c691dbec65a3f1bc36e313c4ef05c4090eb7964f45c9a3a40fba7
  resourceVersion: "358"
  ...
  uid: bad20dc1-27f7-11e6-8991-5254007984a5
```

enoodle commented:
@miq-bot remove_label wip

zeari commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -7,10 +7,21 @@ def ems_inv_to_hashes(inventory)
 
+      def get_openshift_images(inventory)
+        process_collection(inventory["image"], :container_images) { |n| parse_openshift_image(n) }
+        @data[:container_images].uniq! { |n| n[:image_ref] }
```
Could you explain this line? Are you expecting duplicates?

enoodle replied:
Yes because `get_images` from the kubernetes refresh parser will push its images to `@data` and then `process_collection` here will also do that so each image that will be identified by both will be inserted twice (but it will be the same object). So we need to filter them out. In `@data_index` we won't have any duplications because we are just updating images that were added before. Another solution could be to do `@data[:container_images] = []` before the `process_collection` here and I think I will change to it now.

enoodle commented:
@zeari 
1) Yes, I will update soon
2) Only when we will have the pullable image id. After we will have a kubernetes with https://github.com/kubernetes/kubernetes/pull/33014 running underneath. Until then the images will be duplicated.

@miq-bot add_label wip

enoodle commented:
@zeari added UI support. PTAL

enoodle commented:
@miq-bot remove_label wip

zeari commented:
@enoodle 

1) Please export the enviroment variables into another table similar to how it looks for containers.
look in `app/views/container/_container_show.html.haml`
2) can you make `command`, `Entry Point` and `Export ports` looks nicer?

enoodle commented:
@zeari updated, PTAL

zeari commented on app/helpers/container_image_helper/textual_summary.rb:
```diff
@@ -79,5 +79,41 @@ def textual_openscap_failed_rules_high
+        value.nil? ? "REFERENCE" : "VALUE",
+        {:value => value, :expandable => true}
+      ]
+    end
```
I think this already exists for containers. so we can reuse the code from there by moving it to container_summary_helper.

enoodle commented:
@zeari I made some updates according to your comments. PTAL

enoodle commented:
I had to make another change as I found out that images that are found from both methods might not have a name in the docker manifest. I also had to return the `uniq!` because I found out it was necessary. When the `docker-pullable` patch will be merged we will have those images united

![duplicate_images](https://cloud.githubusercontent.com/assets/3123328/19151839/a954ecb0-8bd7-11e6-8372-a4683cdab6f0.png)

simon3z commented on app/models/container_image.rb:
```diff
@@ -6,6 +6,7 @@ class ContainerImage < ApplicationRecord
 
 
   DOCKER_IMAGE_PREFIX = "docker://"
+  DOCKER_PULLABLE_PREFIX = "docker-pullable://"
```
@enoodle is this finalized on the OpenShift side?

enoodle replied:
I am almost sure that it is. The patch that is introducing it to kuberentes was merged but has since been reverted, we will have to see what happens there but there was an agreement on this part

[1] https://github.com/kubernetes/kubernetes/pull/33014/files#diff-0d64fd1b097fff84be9e16633e6acf20R46

simon3z commented on app/models/container_image.rb:
```diff
@@ -45,9 +46,12 @@ def operating_system=(value)
+    [DOCKER_IMAGE_PREFIX, DOCKER_PULLABLE_PREFIX].each do |prefix|
+      if image_ref.start_with?(prefix)
+        return image_ref[prefix.length..-1]
+      end
```
@enoodle you can do a 1-liner here

enoodle replied:
will do.

simon3z commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -109,6 +120,50 @@ def parse_build_pod(build_pod)
+
+      def parse_exposed_ports(exposed_ports)
+        exposed_ports.to_h.keys.each_with_object({}) do |port, h|
+          n, p = port.to_s.split('/')
```
@enoodle do we want to be safe on this one and specify the number of splits we expect? (1)

enoodle replied:
right, I will take care of it.

simon3z commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -109,6 +120,50 @@ def parse_build_pod(build_pod)
+
+      def parse_env_variables(env_variables)
+        env_variables.each_with_object({}) do |var_def, h|
+          name, value = var_def.split('=')
```
@enoodle are you sure it's this simple? Also what if the value contains a `"="`?

enoodle replied:
right, I will take care of it. Variables are always in the form of VARNAME=VALUE even if the value contains `"="` we will be safe if we get the name to be the first part of the split and the value all the rest.

simon3z commented:
@dclarizio can anyone from your side review the UI?

@miq-bot add_label euwe/yes

simon3z commented on app/models/container_image.rb:
```diff
@@ -45,9 +46,10 @@ def operating_system=(value)
-    if image_ref.start_with?(DOCKER_IMAGE_PREFIX)
-      return image_ref[DOCKER_IMAGE_PREFIX.length..-1]
+    [DOCKER_IMAGE_PREFIX, DOCKER_PULLABLE_PREFIX].each do |prefix|
+      return image_ref[prefix.length..-1] if image_ref.start_with?(prefix)
```
@enoodle I think we should add a new `docker_pullable_id` as well (which is returning only if prefix is `docker-pullable://`). Not for this PR though.

Also remember that this part (`docker_pullable_id`) may be very much relevant from the smart state analysis side (cc @moolitayer )

simon3z commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -109,6 +120,50 @@ def parse_build_pod(build_pod)
+        exposed_ports.to_h.keys.each_with_object({}) do |port, h|
+          n, p = port.to_s.split('/', 2)
+          h[n] = p
+        end
```
@enoodle I don't understand this. You create an hash with `protocol => port`? E.g. `{'tcp' => 8080}`? What if there are multiple ports of the same protocol are exposed? Shouldn't this just be a list of exposed ports? `['tcp/8080', 'tcp/80', 'udp/1111' ...]`

enoodle commented:
@zeari, @simon3z This waits review since I added the custom attributes with @zeari. PTAL

zeari commented:
@enoodle This still LGTM but notice that https://github.com/ManageIQ/manageiq/pull/11806/files changes the way we save labels in the parser and in save_inventory)*(`:labels_and_tags =>:labels`)

enoodle commented:
@zeari ok I rebased on it.
@simon3z PTAL, I think this is ready

simon3z commented:
> @simon3z PTAL, I think this is ready

@enoodle please check why travis is failing.

enoodle commented:
@simon3z  ping, This is all green now.

simon3z commented:
@enoodle I am hesitant to merge this now. I am ok to merge it on master but I don't want to backport it to Euwe yet. It could be considered for a z-stream of Euwe.

Bottom line we could merge on master now :+1: and backport this just right after Euwe is done to ensure proper testing for the first z-stream.

cc @bazulay

enoodle commented:
I made some updates to better parse the images after see the new Openshift and docker versions work together.
I also agree that it will be better to merge this to master now and backport it later to z stream after we see that this works well. I don't expect this to break anything if it doesn't work but it might present duplicate images.

enoodle commented:
@miq-bot remove_label euwe/yes
I am removing this label for now according to @simon3z 's comment but we will consider backporting it later.

enoodle commented:
@bazulay @simon3z This already has LGRM from @zeari. Can we merge it to master as suggested above to get more confidence with it? It will only fully work when we have the required fixes to docker and Openshift, but it won't break anything until then and the extra images and data will be available.

simon3z commented:
@enoodle in one of your emails yesterday I saw you mentioned that "there are still some adjustments needed". Can you mention here what they are?

Please decouple the fix needed (that we want to backport to euwe) from the collection of additional information (to be merged/backported later).

enoodle commented:
@simon3z https://github.com/ManageIQ/manageiq/pull/12711

enoodle commented:
@zeari PTAL , added a commit to collect docker_labels and tests.

enoodle commented:
These commits were merged in https://github.com/ManageIQ/manageiq/pull/12851
